[VietAnh1010] iP

VietAnh1010 opened the pull request:
# Voile

> [Click here to see the introduction of _**Voile**_](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

Voile is a chatbot.

- Text-based
- ~~Very~~ easy to use 😄 

Configuring Voile may be a pain. Be careful!

1. Your OS must be able to run GUI
1. You can then run **Voile** using the `java` command

Features (some are being developed right now):

- [x] Add tasks
- [x] List tasks
- [x] Remove tasks
- [x] Mark/unmark tasks
- [x] Find tasks
- [ ] GUI (being developed right now)

You can also develop **Voile** yourself. Start by writing:
```java
public static void main(String[] arg) { ... }
```

Commit:
Add Gradle support

VietAnh1010 committed:
Implement the skeletal version of Duke for Level-1

VietAnh1010 committed:
Add the ability to store and display texts

VietAnh1010 committed:
Add the ability to mark tasks as done

VietAnh1010 committed:
Add support for tracking different types of task

VietAnh1010 committed:
Add support for deleting tasks from list

VietAnh1010 committed:
Add support for VSCode

VietAnh1010 committed:
Setup test using IO redirection

VietAnh1010 committed:
Modify code to better adhere to OOP principles

VietAnh1010 committed:
Save tasks in the hard disk

VietAnh1010 committed:
Store deadline dates as LocalDate objects

VietAnh1010 committed:
Merge branch 'branch-Level-7'

VietAnh1010 committed:
Merge branch 'branch-Level-8'

VietAnh1010 committed:
Divide classes into packages

VietAnh1010 committed:
Update gradle to verson 7.6

VietAnh1010 committed:
Merge branch 'add-gradle-support'

VietAnh1010 committed:
Add some simple JUnit tests

VietAnh1010 committed:
Allow users to find tasks by a keyword

VietAnh1010 committed:
Add pedantic checkstyle

VietAnh1010 committed:
Add JavaDoc to public methods

VietAnh1010 committed:
Merge branch 'branch-A-JavaDoc'

VietAnh1010 committed:
Merge branch 'branch-A-CodingStandard'

VietAnh1010 committed:
Merge branch 'branch-Level-9'

VietAnh1010 committed:
Parser.java: fix typo in JavaDoc comment

VietAnh1010 committed:
Add more types of exception

VietAnh1010 committed:
Add javafx dependencies and modify checkstyle

VietAnh1010 committed:
Remove trailing whitespaces in java files

VietAnh1010 committed:
Add exclude rule for checkstyle

IsabelChong commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.Scanner;
+import java.util.stream.Collectors;
+
+public class UI {
```
Should this class be named "Ui" instead?

IsabelChong commented on src/main/java/duke/DukeUtils.java:
```diff
@@ -0,0 +1,63 @@
+    private static final DateTimeFormatter INPUT_FORMAT =
+            DateTimeFormatter.ofPattern(INPUT_FORMAT_STRING);
+    private static final DateTimeFormatter OUTPUT_FORTMAT =
+            DateTimeFormatter.ofPattern(OUTPUT_FORMAT_STRING);
```
I like how you kept these fixed date formats into a constant 👍🏻

IsabelChong commented on src/main/java/duke/DukeUtils.java:
```diff
@@ -0,0 +1,63 @@
+     *
+     * @param input the input string
+     * @return an {@code Optional} instance holding the conversion result
+     */
```
I like how you have written your javadocs with tags! They are also written clearly so I understand what the methods do. I notice this good practice in several other methods 😊~

IsabelChong commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,206 @@
+public class Parser {
+
+    private enum CommandKeyword {
+
```
Just a genuine question, is this space after the enum declaration a good practice?

VietAnh1010 replied:
This enum is a complex class (values, field and constructor). Adding space after the declaration looks better (at least for me).

IsabelChong commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,131 @@
+        return listTasksFrom(tasks.stream());
+    }
+
+    public String listTasksContainKeyword(String keyword) {
+        return listTasksFrom(tasks.stream().filter(task -> task.contains(keyword)));
```
Should you have javadoc for this method too?

VietAnh1010 replied:
Thanks for reminding me of the Javadoc. I'm quite busy right now and haven't got time to finish all the Javadoc.

IsabelChong commented on src/main/java/duke/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,35 @@
+import duke.UI;
+import duke.task.Task;
+
+public class AddTaskCommand extends Command {
```
Should you have javadoc for this class and the other child classes since you are inheriting from an abstract class?

IsabelChong commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,59 @@
+    private static final Class<DukeRuntimeException> exceptionClass = DukeRuntimeException.class;
+
+    @Test
+    public void shouldCorrectlyParseSomeCommand() {
```
Perhaps you can follow the naming convention of test methods which goes by 
`featureUnderTest_testScenario_expectedBehavior()`
e.g. `sortList_emptyList_exceptionThrown()`,` getMember_memberNotFount_nullReturned()`

IsabelChong reviewed (commented):
It shows from your code that you are experienced and have good practices following the Java Coding Standards! 
Just some nits to fix. Overall, code is user-friendly, and neat 😊.

cjyothika commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,25 @@
+                String input = UI.readCommand();
+                Command cmd = Parser.parseCommand(input);
+                cmd.execute(list);
+                if (cmd.isExit()) {
```
Perhaps you can incorporate this condition into your while loop to avoid nesting?

cjyothika commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,206 @@
+     */
+    public static Command parseCommand(String input) {
+        Parser p = new Parser(input);
+        CommandKeyword kw = p.parseKeyword();
```
Perhaps these variable names can be more descriptive? eg. keyword instead of kw

cjyothika commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,131 @@
+ */
+public class TaskList implements Serializable {
+
+    private static final long serialVersionUID = 1094392670804693665L;
```
Perhaps this constant name can be in all caps?

VietAnh1010 replied:
This is a special constant, used by `java.io.Serializable` to serialize Java object into a byte stream.

cjyothika commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,206 @@
+                cmd = new FindTaskCommand(keyword);
+                break;
+            }
+            default:
```
Good job on having a default branch! It's a common mistake to not include one.

cjyothika reviewed (commented):
Nice work! Your code is readable and concise overall with lots of abstraction. It looks like you're a seasoned programmer. I've given a few minor suggestions.

VietAnh1010 committed:
First, basic version of the GUI

VietAnh1010 committed:
Add FXML, create GUI

VietAnh1010 committed:
Encapsulate the result of commands into a class

VietAnh1010 committed:
Clean up messy GUI code

VietAnh1010 committed:
Group all exceptions into a single package

VietAnh1010 committed:
Restructure code

VietAnh1010 committed:
Fix checkstyle xml files

VietAnh1010 committed:
build.gradle: enable assertions when execute gradle "run" task

Assertions are disabled by default.

Using assertions is a convenient way to ensure invariants when developing the app.

Assertions are enabled only when execute gradle "run" task.

Assertions are not enabled in other tasks, as they are not necessary in these cases.

VietAnh1010 committed:
Add more JavaDoc, following the coding style

VietAnh1010 committed:
TaskList: modify "toString" method

This is a trivial commit.

VietAnh1010 committed:
Merge pull request #1 from VietAnh1010/branch-A-Streams

Complete A-Streams

VietAnh1010 committed:
Merge pull request #2 from VietAnh1010/branch-A-CodeQuality

Complete A-CodeQuality

VietAnh1010 committed:
DukeUtils: add non-null assertions

Static methods in this file assume that inputs are not null.

However, null may be passed into these methods due to programming mistakes. Therefore, assertions
should be added to detect these potential mistakes.

Add assertions that enforce non-null inputs.

VietAnh1010 committed:
Merge pull request #3 from VietAnh1010/branch-A-Assertions

Complete A-Assertions

VietAnh1010 committed:
Add CI, and reformat files

VietAnh1010 committed:
test.yml: Fix setup-java@v3 action

VietAnh1010 committed:
test.yml: try add gradle cache

VietAnh1010 committed:
Add the ability to list unique descriptions

VietAnh1010 committed:
Add another command to list all unique description

VietAnh1010 committed:
Add the ability to sort deadline tasks

VietAnh1010 committed:
Implement multiple CRUD requirements

VietAnh1010 committed:
Implement an applicative parser to parse commands

VietAnh1010 committed:
ApplicativeParser: add a helper cast function for PECS

VietAnh1010 committed:
CommandParserTest: create new file for unit test

VietAnh1010 committed:
Fix unit test

VietAnh1010 committed:
ApplicativeParser: add Javadoc for class ApplicativeParser

VietAnh1010 committed:
Rename some commands

VietAnh1010 committed:
Add help and clear command

VietAnh1010 committed:
Reorganize the repo's structure

VietAnh1010 committed:
Finalize extension

VietAnh1010 committed:
FindCommand: eliminate redudant import

VietAnh1010 committed:
Add User Guide

VietAnh1010 committed:
Finish User Guide

VietAnh1010 committed:
Add JavaDoc for many classes

VietAnh1010 committed:
task: Add more JavaDoc

VietAnh1010 committed:
Improve GUI

VietAnh1010 committed:
Further improve GUI

VietAnh1010 committed:
HelpCommand: write help messages

VietAnh1010 committed:
Update User Guide

VietAnh1010 committed:
Modify some command logic

VietAnh1010 committed:
Duke: rename to Voile

VietAnh1010 committed:
command: add JavaDoc

VietAnh1010 committed:
UnitTest: add more test for ApplicativeParser

VietAnh1010 committed:
Messages: store common messages

VietAnh1010 committed:
CommandParser: modify error messages

VietAnh1010 committed:
Modify response messages

VietAnh1010 committed:
SortCommand: allow grouping and sorting

VietAnh1010 committed:
Add JavaDoc comments for GUI components

VietAnh1010 committed:
Change the title to Voile

VietAnh1010 committed:
style.css: change background color's 'a' attribute

VietAnh1010 committed:
UserGuide: change Duke to Voile

VietAnh1010 committed:
Ui.png: change with a better version

VietAnh1010 committed:
UserGuide: make all 'Voile' bold
